src/cell_list.h

```diff
@@ -0,0 +1,223 @@
+                   current_distance <= range->distance ) {
+                ++current_distance;
+                if( current_distance > distance_to_nearest_cell_wall ) {
+                    load();
+                }
+                current = sorted_elements[current_distance].begin();
+            }
+            if( current_distance > range->distance ) {
+                range = nullptr;
+            }
+            return *this;
+        }
+
+        bool operator==( const cell_list_iterator &rhs ) const {
+            if( range == nullptr && rhs.range == nullptr ) {
+                return true;
+            }
+            if( range != rhs.range ) {
+                return false;
+            }
+            return current == rhs.current;
+        }
+        bool operator!=( const cell_list_iterator &rhs ) const {
+            return !( *this == rhs );
+        }
+
+    private:
+        int calc_distance() {
+            int distance = range->distance;
+            if( min_bounds.x >= 0 ) {
+                distance = std::min( distance, range->origin.x - min_bounds.x );
+            }
+            if( min_bounds.y >= 0 ) {
+                distance = std::min( distance, range->origin.y - min_bounds.y );
+            }
+            if( max_bounds.x < MAXX ) {
+                distance = std::min( distance, max_bounds.x - range->origin.x );
+            }
+            if( max_bounds.y < MAXY ) {
+                distance = std::min( distance, max_bounds.y - range->origin.y );
+            }
+            return distance;
+        }
+        void add( const std::vector<std::pair<point, T>> &cells ) {
+            for( const std::pair<point, T> &candidate : cells ) {
+                add( candidate );
+            }
+        }
+        void add( const std::pair<point, T> &candidate ) {
+            const int distance = rl_dist( candidate.first, range->origin );
+            if( distance > range->distance ) {
+                return;
+            }
+            sorted_elements[distance].push_back( candidate );
+        }
+        void load() {
+            const point rank_offset{ CELL_SIZE, CELL_SIZE };
+            min_bounds -= rank_offset;
+            max_bounds += rank_offset;
+            distance_to_nearest_cell_wall = calc_distance();
+            cell_ranks_loaded++;
+            const point load_offset( cell_ranks_loaded, cell_ranks_loaded );
+            const point min_cell = range->origin_cell - load_offset;
+            const point max_cell = range->origin_cell + load_offset;
+            const bool min_x_inbounds = min_cell.x >= 0;
+            const bool min_y_inbounds = min_cell.y >= 0;
+            const bool max_x_inbounds = max_cell.x < divide_round_up( MAXX, CELL_SIZE );
+            const bool max_y_inbounds = max_cell.y < divide_round_up( MAXY, CELL_SIZE );
+            // Draw a box...
+            for( int x = min_cell.x; x <= max_cell.x; ++x ) {
+                if( x < 0 || x > MAXX / CELL_SIZE ) {
```

Comment:
```suggestion
                if( x < 0 || x > divide_round_up( MAXX, CELL_SIZE ) ) {
```
But I think it might be even cleaner to add a `static constexpr point cell_dim()` member to `cell_list` and use that everywhere you need to know the dimensions.

Reply:
yes, but num_cells_x() and num_cells_y()